This project has 2 case studies.The first case study (described below) covers concepts taught in Part 1 (first 8 hours
of Neural networks basics).
 
1st case study - Project 1:
 
The case study is from an open source dataset from Kaggle. 

Link to the Kaggle project site:
https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling

Given a Bank customer, can we build a classifier which can determine whether they will leave or
not using Neural networks?
 
Case file: 

bank.csv
 
The points distribution for this case is as follows:
1. Read the dataset in a new python notebook.
2. Drop the columns which are unique for all users like IDs
3. Distinguish the feature and target set
4. Divide the data set into Train and test sets
5. Normalize the train and test data
6. Initialize &amp; build the model
7. Optimize the model
9. Predict the results using 0.5 as a threshold
10. Print the Accuracy score and confusion matrix

Read the dataset in a new python notebook.

In [380]:
%tensorflow_version 2.0
import tensorflow as tf
print(tf.__version__)

`%tensorflow_version` only switches the major version: `1.x` or `2.x`.
You set: `2.0`. This will be interpreted as: `2.x`.


TensorFlow is already loaded. Please restart the runtime to change versions.
2.1.0-rc1


In [0]:
import pandas as pd
import numpy as np

In [382]:
from google.colab import drive

drive.mount('/content/gdrive')
root_path = "gdrive/My Drive/GreatLearning/myprojects/Residency6/bank.csv"  #change dir to your project folder
bankdf = pd.read_csv(root_path)

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [383]:
bankdf.head(10)

RowNumber  CustomerId   Surname  ...  IsActiveMember EstimatedSalary Exited
0          1    15634602  Hargrave  ...               1       101348.88      1
1          2    15647311      Hill  ...               1       112542.58      0
2          3    15619304      Onio  ...               0       113931.57      1
3          4    15701354      Boni  ...               0        93826.63      0
4          5    15737888  Mitchell  ...               1        79084.10      0
5          6    15574012       Chu  ...               0       149756.71      1
6          7    15592531  Bartlett  ...               1        10062.80      0
7          8    15656148    Obinna  ...               0       119346.88      1
8          9    15792365        He  ...               1        74940.50      0
9         10    15592389        H?  ...               1        71725.73      0

[10 rows x 14 columns]

In [384]:
bankdf.shape

(10000, 14)

In [385]:
bankdf.columns

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited'],
      dtype='object')

In [386]:
bankdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
RowNumber          10000 non-null int64
CustomerId         10000 non-null int64
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             10000 non-null int64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


There are no null values.
Three categorical type variables (Surname,Geography and Gender)

Drop the columns which are unique for all users like IDs.

Distinguish the feature and target set.

In [388]:
#importing the dataset
dataset = bankdf
X = dataset.iloc[:,3:13].values # Credit Score through Estimated Salary
y = dataset.iloc[:, 13].values # Exited
print(dataset.shape)
print(X.shape)
print(y.shape)

(10000, 14)
(10000, 10)
(10000,)


In [389]:
#checking unique values for categorical columns
print('Unique values in Geography column: ', dataset.Geography.unique())
print('Unique values in Gender column: ', dataset.Gender.unique())
print('Unique values in Exited column: ', dataset.Exited.unique())

Unique values in Geography column:  ['France' 'Spain' 'Germany']
Unique values in Gender column:  ['Female' 'Male']
Unique values in Exited column:  [1 0]


In [353]:
# Encoding categorical data into scalar values.
# Geography column has 3 unique values : France, Spain and Germany
print('Geography values before encoding: ', X[:10,1])
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
label_X_geography_encoder = LabelEncoder()
X[:,1] = label_X_geography_encoder.fit_transform(X[:,1])
print('Geography values after encoding: ', X[:10,1])

Geography values before encoding:  ['France' 'Spain' 'France' 'France' 'Spain' 'Spain' 'France' 'Germany'
 'France' 'France']
Geography values after encoding:  [0 2 0 0 2 2 0 1 0 0]


In [390]:
# Encoding categorical data into scalar values.
# Gender column has 2 unique values : Male and Female
print('Gender values before encoding: ', X[:10,2])
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
label_X_gender_encoder = LabelEncoder()
X[:,2] = label_X_gender_encoder.fit_transform(X[:,2])
print('Gender values after encoding: ', X[:10,2])

Gender values before encoding:  ['Female' 'Female' 'Female' 'Female' 'Female' 'Male' 'Male' 'Female'
 'Male' 'Male']
Gender values after encoding:  [0 0 0 0 0 1 1 0 1 1]


In [0]:
# Converting the string features into their own dimensions. 
# Gender doesn't matter here because its binary
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
ct = ColumnTransformer([("Geography", OneHotEncoder(), [1])], remainder = 'passthrough')
X = ct.fit_transform(X) 

In [0]:
X = X[:,1:]

Divide the data set into Train and test sets.

In [0]:
# Splitting the dataset into the Training and Testing set.
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.2, random_state = 9)

Normalize the train and test data

In [0]:
# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [395]:
print(f'Training shapes: {X_train.shape}, {y_train.shape}')
print(f'Testing shapes: {X_test.shape}, {y_test.shape}')

Training shapes: (8000, 11), (8000,)
Testing shapes: (2000, 11), (2000,)


Initialize & build the model

In [0]:
#Importing required keras modules
from tensorflow.keras.models import Sequential
from tensorflow.keras.backend import clear_session
from tensorflow.keras.layers import Reshape, Dense, Dropout, BatchNormalization, Activation

In [0]:
clear_session()
# Initializing the DNN model
model1 = Sequential()

# #Add Input Layer (first hidden layer)
model1.add(Dense(units = 6, input_dim = X_train.shape[1], activation = 'relu'))

# Add second hidden layer
model1.add(Dense(units=6, activation = 'relu')) 

# Adding the output layer
# We use the sigmoid to get probability outcomes for binary class
model1.add(Dense(units=1, activation = 'sigmoid'))

model1.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])

In [398]:
model1.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 6)                 72        
_________________________________________________________________
dense_1 (Dense)              (None, 6)                 42        
_________________________________________________________________
dense_2 (Dense)              (None, 1)                 7         
Total params: 121
Trainable params: 121
Non-trainable params: 0
_________________________________________________________________


From the model summary:
Model built above has three layers.Input layer, hidden layer and output layer with output layer has one neuron for binary class output

In [399]:
model1.fit(X_train, y_train, validation_data=(X_test, y_test), batch_size=32, epochs=10)

Train on 8000 samples, validate on 2000 samples
Epoch 1/10
8000/8000 [==============================] - 1s 90us/sample - loss: 0.5562 - accuracy: 0.7701 - val_loss: 0.5074 - val_accuracy: 0.7885
Epoch 2/10
8000/8000 [==============================] - 0s 50us/sample - loss: 0.4785 - accuracy: 0.7976 - val_loss: 0.4719 - val_accuracy: 0.7885
Epoch 3/10
8000/8000 [==============================] - 0s 51us/sample - loss: 0.4493 - accuracy: 0.7974 - val_loss: 0.4479 - val_accuracy: 0.7955
Epoch 4/10
8000/8000 [==============================] - 0s 52us/sample - loss: 0.4256 - accuracy: 0.8154 - val_loss: 0.4263 - val_accuracy: 0.8210
Epoch 5/10
8000/8000 [==============================] - 0s 52us/sample - loss: 0.4067 - accuracy: 0.8264 - val_loss: 0.4112 - val_accuracy: 0.8345
Epoch 6/10
8000/8000 [==============================] - 0s 52us/sample - loss: 0.3913 - accuracy: 0.8351 - val_loss: 0.3970 - val_accuracy: 0.8430
Epoch 7/10
8000/8000 [==============================] - 0s 50us/sample

Optimize the model

For Optimizing model, following can be attempted,

1.Number of hidden layer and number of neurons in each hidden layer can be adjusted.

2.Batch normalization layer can be added.

3.Dropout layer can be added

3.selection of optimizer

4.value initialization

5.parameter tuning for selected optimizer


Following changes will be done in above model to optimize:

1.One hidden layer will added

2.Number of neurons will be increased to 8 from 6

3.Batch Normalization will be added after each layer

4.Number of epochs will be increased to 100 from 10

5.Batch size will be reduced to 10 from 32

6.Dropout layer will be added with 10% drop after each hidden layer

In [0]:
clear_session()

# Initialize Sequential model
model2 = Sequential()

#Add Input Layer (first hidden layer)
model2.add(Dense(units = 8, activation = 'relu', input_dim = X_train.shape[1]))

# BatchNormalization layer
model2.add(BatchNormalization())
# Dropout layer
model2.add(Dropout(rate=0.1))

#Add Second hidden layer
model2.add(Dense(units = 8, activation = 'relu'))

# BatchNormalization layer
model2.add(BatchNormalization())
# Dropout layer
model2.add(Dropout(rate=0.1))

#Add Third hidden layer
model2.add(Dense(units = 8, activation = 'relu'))

# BatchNormalization layer
model2.add(BatchNormalization())
# Dropout layer
model2.add(Dropout(rate=0.1))

# Add Output Layer
model2.add(Dense(1, activation='sigmoid'))

In [0]:
model2.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])

In [402]:
model2.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 8)                 96        
_________________________________________________________________
batch_normalization (BatchNo (None, 8)                 32        
_________________________________________________________________
dropout (Dropout)            (None, 8)                 0         
_________________________________________________________________
dense_1 (Dense)              (None, 8)                 72        
_________________________________________________________________
batch_normalization_1 (Batch (None, 8)                 32        
_________________________________________________________________
dropout_1 (Dropout)          (None, 8)                 0         
_________________________________________________________________
dense_2 (Dense)              (None, 8)                 7

In [403]:
model2.fit(X_train, y_train, validation_data=(X_test, y_test), batch_size=10, epochs=100)

Train on 8000 samples, validate on 2000 samples
Epoch 1/100
8000/8000 [==============================] - 2s 303us/sample - loss: 0.5673 - accuracy: 0.7297 - val_loss: 0.4925 - val_accuracy: 0.7905
Epoch 2/100
8000/8000 [==============================] - 2s 199us/sample - loss: 0.4972 - accuracy: 0.7906 - val_loss: 0.4718 - val_accuracy: 0.7940
Epoch 3/100
8000/8000 [==============================] - 2s 202us/sample - loss: 0.4769 - accuracy: 0.7959 - val_loss: 0.4523 - val_accuracy: 0.7970
Epoch 4/100
8000/8000 [==============================] - 2s 200us/sample - loss: 0.4621 - accuracy: 0.7983 - val_loss: 0.4418 - val_accuracy: 0.8015
Epoch 5/100
8000/8000 [==============================] - 2s 202us/sample - loss: 0.4487 - accuracy: 0.8011 - val_loss: 0.4337 - val_accuracy: 0.8110
Epoch 6/100
8000/8000 [==============================] - 2s 201us/sample - loss: 0.4420 - accuracy: 0.8031 - val_loss: 0.4271 - val_accuracy: 0.8155
Epoch 7/100
8000/8000 [==============================] - 2

Removing Dropout layers to check the model performance.

Number of neurons in each layer as 7.

Adding one more hidden layer

In [405]:
clear_session()

# Initialize Sequential model
model3 = Sequential()

#Add Input Layer (first hidden layer)
model3.add(Dense(units = 7, activation = 'relu', input_dim = X_train.shape[1]))

# BatchNormalization layer
model3.add(BatchNormalization())

#Add Second hidden layer
model3.add(Dense(units = 7, activation = 'relu'))

# BatchNormalization layer
model3.add(BatchNormalization())

#Add Third hidden layer
model3.add(Dense(units = 7, activation = 'relu'))

# BatchNormalization layer
model3.add(BatchNormalization())

#Add Fourth hidden layer
model3.add(Dense(units = 7, activation = 'relu'))

# BatchNormalization layer
model3.add(BatchNormalization())

# Add Output Layer
model3.add(Dense(1, activation='sigmoid'))

model3.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])

model3.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 7)                 84        
_________________________________________________________________
batch_normalization (BatchNo (None, 7)                 28        
_________________________________________________________________
dense_1 (Dense)              (None, 7)                 56        
_________________________________________________________________
batch_normalization_1 (Batch (None, 7)                 28        
_________________________________________________________________
dense_2 (Dense)              (None, 7)                 56        
_________________________________________________________________
batch_normalization_2 (Batch (None, 7)                 28        
_________________________________________________________________
dense_3 (Dense)              (None, 7)                 5

In [406]:
model3.fit(X_train, y_train, validation_data=(X_test, y_test), batch_size=10, epochs=100)

Train on 8000 samples, validate on 2000 samples
Epoch 1/100
8000/8000 [==============================] - 3s 354us/sample - loss: 0.5296 - accuracy: 0.7521 - val_loss: 0.4697 - val_accuracy: 0.7965
Epoch 2/100
8000/8000 [==============================] - 2s 234us/sample - loss: 0.4490 - accuracy: 0.8049 - val_loss: 0.4411 - val_accuracy: 0.8115
Epoch 3/100
8000/8000 [==============================] - 2s 235us/sample - loss: 0.4336 - accuracy: 0.8066 - val_loss: 0.4311 - val_accuracy: 0.8155
Epoch 4/100
8000/8000 [==============================] - 2s 233us/sample - loss: 0.4218 - accuracy: 0.8069 - val_loss: 0.4264 - val_accuracy: 0.8160
Epoch 5/100
8000/8000 [==============================] - 2s 234us/sample - loss: 0.4152 - accuracy: 0.8087 - val_loss: 0.4150 - val_accuracy: 0.8280
Epoch 6/100
8000/8000 [==============================] - 2s 237us/sample - loss: 0.4119 - accuracy: 0.8148 - val_loss: 0.4039 - val_accuracy: 0.8350
Epoch 7/100
8000/8000 [==============================] - 2

With Adam optimizer, once the initial learning rate is set, there is no need to tune the learning rate.

First and Second models above used adam optimizer.

Building model4 with SGD with neserov momentum with same epoch and batchsize, to compare the effect of optimizer change.

In [0]:
clear_session()

# Initialize Sequential model
model4 = Sequential()

#Add Input Layer (first hidden layer)
model4.add(Dense(units = 7, activation = 'relu', input_dim = X_train.shape[1]))

# BatchNormalization layer
model4.add(BatchNormalization())
# Dropout layer
model4.add(Dropout(rate=0.1))

#Add Second hidden layer
model4.add(Dense(units = 7, activation = 'relu'))

# BatchNormalization layer
model4.add(BatchNormalization())
# Dropout layer
model4.add(Dropout(rate=0.1))

#Add Third hidden layer
model4.add(Dense(units = 7, activation = 'relu'))

# BatchNormalization layer
model4.add(BatchNormalization())
# Dropout layer
model4.add(Dropout(rate=0.1))

# Add Output Layer
model4.add(Dense(1, activation='sigmoid'))

#Compile the model
sgd_optimizer = tf.keras.optimizers.SGD(lr=0.01, momentum = 0.9, decay=0.001,nesterov = True)
model4.compile(optimizer=sgd_optimizer, loss = 'binary_crossentropy', metrics=['accuracy'])

In [408]:
model4.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 7)                 84        
_________________________________________________________________
batch_normalization (BatchNo (None, 7)                 28        
_________________________________________________________________
dropout (Dropout)            (None, 7)                 0         
_________________________________________________________________
dense_1 (Dense)              (None, 7)                 56        
_________________________________________________________________
batch_normalization_1 (Batch (None, 7)                 28        
_________________________________________________________________
dropout_1 (Dropout)          (None, 7)                 0         
_________________________________________________________________
dense_2 (Dense)              (None, 7)                 5

In [409]:
model4.fit(X_train, y_train, validation_data=(X_test, y_test), batch_size=10, epochs=100)

Train on 8000 samples, validate on 2000 samples
Epoch 1/100
8000/8000 [==============================] - 2s 294us/sample - loss: 0.5036 - accuracy: 0.7912 - val_loss: 0.4751 - val_accuracy: 0.7890
Epoch 2/100
8000/8000 [==============================] - 2s 202us/sample - loss: 0.4572 - accuracy: 0.7960 - val_loss: 0.4504 - val_accuracy: 0.7935
Epoch 3/100
8000/8000 [==============================] - 2s 201us/sample - loss: 0.4437 - accuracy: 0.8008 - val_loss: 0.4420 - val_accuracy: 0.8065
Epoch 4/100
8000/8000 [==============================] - 2s 208us/sample - loss: 0.4411 - accuracy: 0.8043 - val_loss: 0.4390 - val_accuracy: 0.8110
Epoch 5/100
8000/8000 [==============================] - 2s 212us/sample - loss: 0.4410 - accuracy: 0.7979 - val_loss: 0.4362 - val_accuracy: 0.8030
Epoch 6/100
8000/8000 [==============================] - 2s 213us/sample - loss: 0.4401 - accuracy: 0.8021 - val_loss: 0.4352 - val_accuracy: 0.8060
Epoch 7/100
8000/8000 [==============================] - 2

From the above observations:

1.Adam optimizer doesnot need learning rate tuning.

2.Given that SGD with nesterov momentum doesnot provide significant difference in the validation accuracy.Adam seems to be preferred for given dataset.

3.Model gives better test accuracy with added hidden layer and increasing neurons in some layers.

4.Adding Dropout layer in this model with the given dataset doesnot seem to have much benefit.

5.Calculating further evaluation metrics and prediction based on model3 above which is built with BatchNormalization, without Dropout layer, adam optimizer and activation function as RELU.

Predict the results using 0.5 as a threshold

In [410]:
y_pred = model3.predict(X_test)
print(y_pred)

[[0.01401042]
 [0.31011286]
 [0.25074044]
 ...
 [0.07809349]
 [0.2024738 ]
 [0.12804799]]


In [0]:
y_pred = y_pred > 0.5

Print the Accuracy score and confusion matrix

In [412]:
from sklearn.metrics import confusion_matrix, classification_report
print('Classification Report:')
cr = classification_report(y_test,y_pred)
print(cr)

print('Confusion Matrix:')
cm = confusion_matrix(y_test,y_pred)
print(cm)

Classification Report:
              precision    recall  f1-score   support

           0       0.88      0.95      0.91      1578
           1       0.73      0.50      0.59       422

    accuracy                           0.85      2000
   macro avg       0.80      0.72      0.75      2000
weighted avg       0.84      0.85      0.84      2000

Confusion Matrix:
[[1499   79]
 [ 212  210]]


In [413]:
print('Model Evaluation on test data: ')
results = model3.evaluate(X_test, y_test, verbose=0)
print('Test loss, Test Accuracy : ', results)

Model Evaluation on test data: 
Test loss, Test Accuracy :  [0.3514164969921112, 0.8545]


Model3 trained with 100 epochs and batch size of 10 and adam optimizers with BatchNormalization gives the test accuracy of around 85% and loss (binary crossentropy) value of around 0.351

Model summary in each model building step above depicts the model representation.

Batch Normalization layer normalizes the previous layer output at each batch.This leads to non-trainable parameters mu and signma which are not updated by gradient descent and trainable parameters gamma and beta which are updated based on required scaling and shifting as per gradient descent.Batchnormalization layer is used to reduce overfitting and also normalization helps reduce the effect of varying distribution of input data when test data is fed to the model.

Dropout layer helps to reduce overfitting.In abover models, Dropout layer doesnot seem to have much effect on validation accuracy improvement.


Three models were built above to optimize first model which was very basic with on input layer, one hidden layer and one output layer.

Then additional hidden layer, BatchNormatization and Dropout layer was added in first model to optimize the model.Epoch was reduced and Epoch was increased.Optimizer was adam.

Third model removed Dropout layers in second model to check the effect on validation accuracy for this given dataset.

Fourth model used SGD with nesterov momentum to check the effect of chanding the optimizer for this given dataset.

Finally model3 was used for prediction and metrics evaluation.